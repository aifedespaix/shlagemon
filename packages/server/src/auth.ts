import type { PrismaClient } from '@prisma/client'
import type { Redis } from 'ioredis'
import process from 'node:process'
import jwt from 'jsonwebtoken'
import { generateId } from 'lucia'

/**
 * Short-lived JWT session handling using Lucia-generated identifiers.
 * Sessions are persisted in Redis with a strict TTL.
 */
export interface AuthContext {
  redis: Redis
  prisma: PrismaClient
}

const JWT_SECRET = process.env.JWT_SECRET ?? 'change-me'
const SESSION_TTL_SECONDS = 60 // 1 minute

export async function issueJwt(ctx: AuthContext, userId: string): Promise<string> {
  const sessionId = generateId(32)
  await ctx.redis.set(`session:${sessionId}`, userId, 'EX', SESSION_TTL_SECONDS)
  return jwt.sign({ sub: userId, sid: sessionId }, JWT_SECRET, { expiresIn: SESSION_TTL_SECONDS })
}

export async function validateJwt(ctx: AuthContext, token: string): Promise<{ userId: string } | null> {
  try {
    const decoded = jwt.verify(token, JWT_SECRET) as { sub: string, sid: string }
    const stored = await ctx.redis.get(`session:${decoded.sid}`)
    if (stored !== decoded.sub)
      return null
    return { userId: decoded.sub }
  }
  catch {
    return null
  }
}
