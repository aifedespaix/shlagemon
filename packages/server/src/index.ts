import process from 'node:process'
import { createServer } from './server.js'

const port = Number(process.env.PORT) || 3000

createServer().then(({ fastify }) => {
  fastify.listen({ port, host: '0.0.0.0' }).catch((err) => {
    fastify.log.error(err)
    process.exit(1)
  })
})
