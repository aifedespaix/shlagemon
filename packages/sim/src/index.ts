export * from './engine'
export * from './runScenario'
