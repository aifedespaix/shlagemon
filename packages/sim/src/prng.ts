/**
 * Seedable pseudo-random number generator based on Mulberry32 algorithm.
 * Provides deterministic sequences for reproducible simulations.
 */
export class PRNG {
  private state: number

  constructor(seed: number) {
    if (!Number.isInteger(seed))
      throw new Error('Seed must be an integer')
    this.state = seed >>> 0
  }

  /**
   * Generates the next pseudo-random number in [0, 1).
   */
  next(): number {
    // Mulberry32 algorithm
    let t = this.state += 0x6D2B79F5
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }

  /**
   * Returns current internal state for snapshotting.
   */
  getState(): number {
    return this.state >>> 0
  }

  /**
   * Restores the internal state from snapshot.
   */
  setState(state: number): void {
    if (!Number.isInteger(state))
      throw new Error('State must be an integer')
    this.state = state >>> 0
  }
}
