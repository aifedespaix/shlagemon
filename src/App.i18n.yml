fr:
  title: Shlagémon - Ça sent très fort
  description: Attrape tous les Shlagémons dans ce jeu parodique hilarant ! Collectionne des monstres absurdes, mène des combats délirants et sauve la planète d’une invasion complètement folle. Télécharge maintenant et deviens le meilleur shlagidresseur !
  author: Aife
en:
  title: Shlagemon - It smells very strong
  description: Catch all the Shlagémons in this hilarious parody game! Collect absurd monsters, engage in wacky battles, and save the planet from a totally crazy invasion. Download now and become the ultimate Shlagitrainer!
  author: Aife
