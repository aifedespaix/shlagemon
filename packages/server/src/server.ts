import type { AuthContext } from './auth.js'
import { readFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import process from 'node:process'
import { fileURLToPath } from 'node:url'
import cors from '@fastify/cors'
import { PrismaClient } from '@prisma/client'
import { Server } from 'colyseus'
import Fastify from 'fastify'
import { Redis } from 'ioredis'
import { issueJwt, validateJwt } from './auth.js'
import { GameRoom } from './rooms/GameRoom.js'

const __dirname = dirname(fileURLToPath(import.meta.url))

/**
 * Create a Fastify server instance with Colyseus rooms and JWT authentication.
 */
export async function createServer() {
  const fastify = Fastify({ logger: true })
  await fastify.register(cors)

  const redis = process.env.NODE_ENV === 'test'
  // eslint-disable-next-line new-cap
    ? new (await import('ioredis-mock')).default()
    : new Redis(process.env.REDIS_URL ?? 'redis://localhost:6379')
  const prisma = new PrismaClient()
  const ctx: AuthContext = { redis, prisma }

  // Routes
  fastify.get('/healthz', async () => ({ status: 'ok' }))
  fastify.get('/time', async () => ({ time: Date.now() }))
  fastify.get('/version', async () => {
    const pkg = JSON.parse(readFileSync(join(__dirname, '..', 'package.json'), 'utf8'))
    return { version: pkg.version }
  })

  fastify.post<{ Body: { userId: string } }>('/auth/token', async (req, reply) => {
    const { userId } = req.body
    const user = await prisma.user.findUnique({ where: { id: userId } })
    if (!user) {
      reply.code(401)
      return { error: 'invalid user' }
    }
    const token = await issueJwt(ctx, userId)
    return { token }
  })

  fastify.get('/auth/validate', async (req, reply) => {
    const authHeader = req.headers.authorization
    if (!authHeader) {
      reply.code(401)
      return { error: 'missing token' }
    }
    const [, token] = authHeader.split(' ')
    const result = await validateJwt(ctx, token)
    if (!result) {
      reply.code(401)
      return { error: 'invalid token' }
    }
    return { userId: result.userId }
  })

  // Colyseus + Fastify integration
  const gameServer = new Server({ server: fastify.server })
  gameServer.define('game', GameRoom, { ctx })

  return { fastify, gameServer, ctx }
}
