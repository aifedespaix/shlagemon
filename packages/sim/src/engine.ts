import { PRNG } from './prng'

/**
 * Snapshot of the simulation state.
 */
export interface SimulationSnapshot {
  /** Number of completed simulation ticks. */
  tick: number
  /** Accumulated value influenced by inputs and randomness. */
  value: number
  /** Internal state of the pseudo-random number generator. */
  rngState: number
}

/**
 * Deterministic headless simulation engine with a fixed time step.
 */
export class Engine {
  /** Duration of a single simulation step in milliseconds. */
  readonly stepMs: number

  private accumulator = 0
  private prng: PRNG
  private pendingInput = 0

  /**
   * Mutable simulation state.
   */
  private state: SimulationSnapshot

  constructor(seed: number, stepHz = 20) {
    if (stepHz <= 0)
      throw new Error('Step frequency must be positive')
    this.stepMs = Math.floor(1000 / stepHz)
    this.prng = new PRNG(seed)
    this.state = {
      tick: 0,
      value: 0,
      rngState: this.prng.getState(),
    }
  }

  /**
   * Queues an input to be processed on the next simulation step.
   */
  enqueueInput(amount: number): void {
    if (!Number.isFinite(amount))
      throw new Error('Input must be a finite number')
    this.pendingInput += amount
  }

  /**
   * Advances the simulation by the provided time delta.
   */
  update(deltaMs: number): void {
    if (deltaMs < 0)
      throw new Error('Delta time cannot be negative')
    this.accumulator += deltaMs
    while (this.accumulator >= this.stepMs) {
      this.step()
      this.accumulator -= this.stepMs
    }
  }

  private step(): void {
    const randomContribution = Math.floor(this.prng.next() * 10)
    this.state.tick += 1
    this.state.value += this.pendingInput + randomContribution
    this.pendingInput = 0
    this.state.rngState = this.prng.getState()
  }

  /**
   * Creates a serializable snapshot of the current state.
   */
  getSnapshot(): SimulationSnapshot {
    return { ...this.state }
  }

  /**
   * Restores the simulation state from a snapshot.
   */
  setSnapshot(snapshot: SimulationSnapshot): void {
    this.prng.setState(snapshot.rngState)
    this.state = { ...snapshot }
    this.accumulator = 0
    this.pendingInput = 0
  }
}
