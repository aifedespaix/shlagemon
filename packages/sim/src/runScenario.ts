import type { SimulationSnapshot } from './engine'
import { Engine } from './engine'

/**
 * Input applied at a specific simulation frame.
 */
export interface ScenarioInput {
  /** Frame number at which the input is applied. */
  frame: number
  /** Numeric value added to the simulation on that frame. */
  value: number
}

/**
 * Runs a deterministic scenario for testing purposes.
 * @param seed Seed for the pseudo-random number generator.
 * @param inputs Time-ordered list of frame-based inputs.
 * @returns Array of snapshots taken after each frame.
 */
export function runScenario(seed: number, inputs: ScenarioInput[]): SimulationSnapshot[] {
  const engine = new Engine(seed)
  const ordered = [...inputs].sort((a, b) => a.frame - b.frame)
  const snapshots: SimulationSnapshot[] = []

  const lastFrame = ordered.length ? ordered[ordered.length - 1].frame : 0
  const totalFrames = Math.max(lastFrame + 1, 10)

  let inputIndex = 0
  for (let frame = 0; frame < totalFrames; frame++) {
    while (inputIndex < ordered.length && ordered[inputIndex].frame === frame) {
      engine.enqueueInput(ordered[inputIndex].value)
      inputIndex += 1
    }
    engine.update(engine.stepMs)
    snapshots.push(engine.getSnapshot())
  }

  return snapshots
}
