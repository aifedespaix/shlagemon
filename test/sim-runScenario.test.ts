import { describe, expect, it } from 'vitest'
import { Engine, runScenario } from '../packages/sim/src'

describe('deterministic simulation', () => {
  it('produces stable snapshots', () => {
    const snapshots = runScenario(123, [
      { frame: 2, value: 5 },
      { frame: 5, value: -3 },
    ])
    expect(snapshots).toMatchSnapshot()
  })

  it('restores from snapshot', () => {
    const engine = new Engine(42)
    engine.enqueueInput(3)
    engine.update(engine.stepMs)
    const snap = engine.getSnapshot()

    const restored = new Engine(0)
    restored.setSnapshot(snap)
    expect(restored.getSnapshot()).toEqual(snap)
  })
})
