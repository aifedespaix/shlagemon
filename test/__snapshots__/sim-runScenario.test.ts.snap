// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deterministic simulation > produces stable snapshots 1`] = `
[
  {
    "rngState": 1831565936,
    "tick": 1,
    "value": 7,
  },
  {
    "rngState": 3663131749,
    "tick": 2,
    "value": 8,
  },
  {
    "rngState": 1199730266,
    "tick": 3,
    "value": 17,
  },
  {
    "rngState": 3031296079,
    "tick": 4,
    "value": 19,
  },
  {
    "rngState": 567894596,
    "tick": 5,
    "value": 22,
  },
  {
    "rngState": 2399460409,
    "tick": 6,
    "value": 27,
  },
  {
    "rngState": 4231026222,
    "tick": 7,
    "value": 35,
  },
  {
    "rngState": 1767624739,
    "tick": 8,
    "value": 44,
  },
  {
    "rngState": 3599190552,
    "tick": 9,
    "value": 51,
  },
  {
    "rngState": 1135789069,
    "tick": 10,
    "value": 55,
  },
]
`;
